Python/Product/Analysis/Interpreter/Ast/AstPythonInterpreter.cs

```diff
@@ -0,0 +1,128 @@
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.PythonTools.Analysis;
+using Microsoft.PythonTools.Infrastructure;
+
+namespace Microsoft.PythonTools.Interpreter.Ast {
+    class AstPythonInterpreter : IPythonInterpreter, IModuleContext {
+        private readonly AstPythonInterpreterFactory _factory;
+        private readonly Dictionary<BuiltinTypeId, IPythonType> _builtinTypes;
+        private IBuiltinPythonModule _builtinModule;
+        private readonly ConcurrentDictionary<string, IPythonModule> _modules;
+
+        public AstPythonInterpreter(AstPythonInterpreterFactory factory) {
+            if (factory == null) {
+                throw new ArgumentNullException(nameof(factory));
+            }
+            _factory = factory;
+            _factory.ImportableModulesChanged += Factory_ImportableModulesChanged;
+            _modules = new ConcurrentDictionary<string, IPythonModule>();
+            _builtinTypes = new Dictionary<BuiltinTypeId, IPythonType>();
+        }
+
+        public void Dispose() {
+            _factory.ImportableModulesChanged -= Factory_ImportableModulesChanged;
+        }
+
+        private void Factory_ImportableModulesChanged(object sender, EventArgs e) {
+            ModuleNamesChanged?.Invoke(this, EventArgs.Empty);
+        }
+
+        public event EventHandler ModuleNamesChanged;
+
+        public IModuleContext CreateModuleContext() => this;
+        public IPythonInterpreterFactory Factory => _factory;
+
+        public IPythonType GetBuiltinType(BuiltinTypeId id) {
+            IPythonType res;
+            lock (_builtinTypes) {
+                if (!_builtinTypes.TryGetValue(id, out res)) {
+                    _builtinTypes[id] = res =
+                        _builtinModule?.GetAnyMember(SharedDatabaseState.GetBuiltinTypeName(id, _factory.Configuration.Version)) as IPythonType ??
+                        new AstPythonType(id.ToString());
+                }
+            }
+            return res;
+        }
+
+        public IList<string> GetModuleNames() {
+            return _factory.GetImportableModules().Keys.ToArray();
+        }
+
+        public IPythonModule ImportModule(string name) {
+            IPythonModule mod;
```

Comment:
validate name for null and empty string (there's a call to name.IndexOf which will crash if null)